src/main/java/uk/gov/ons/ctp/integration/rhsvc/event/impl/EventBuilder.java

```diff
@@ -0,0 +1,75 @@
+package uk.gov.ons.ctp.integration.rhsvc.event.impl;
+
+import java.util.Date;
+import java.util.UUID;
+import lombok.Getter;
+import uk.gov.ons.ctp.common.error.CTPException;
+import uk.gov.ons.ctp.common.event.model.GenericEvent;
+import uk.gov.ons.ctp.common.event.model.Header;
+import uk.gov.ons.ctp.common.event.model.RespondentAuthenticatedEvent;
+import uk.gov.ons.ctp.common.event.model.RespondentAuthenticatedResponse;
+import uk.gov.ons.ctp.common.event.model.SurveyLaunchedEvent;
+import uk.gov.ons.ctp.common.event.model.SurveyLaunchedResponse;
+
+public class EventBuilder {
+
+  @Getter
+  public enum EventType {
+    SURVEY_LAUNCHED("RESPONDENT_HOME", "RH"),
+    RESPONDENT_AUTHENTICATED("RESPONDENT_HOME", "RH");
+
+    private String source;
+    private String channel;
+
+    EventType(String source, String channel) {
+      this.source = source;
+      this.channel = channel;
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  public static <T extends GenericEvent, S> T buildEvent(EventType type, S payload)
```

Comment:
This code would be simpler if it was implemented as 2 distinct methods. All the conditional logic and associated exception could then be removed. 
There are only 2 callers to the method, and both have fixed arguments, so the flexibility provided by taking generic arguments is not currently needed. YAGNI !

Reply:
Could use overloaded methods in a Builder type implementation. Would remove the need for Generics and a Switch statement. The idea was as you suggest the buildEvent method could be easily extended in future for all events keeping the same logic each time without having to add a new method every time. As you say are we going to need it?

Reply:
Simplified Eventbuilder as two distinct methods removing generics and associated exceptions as suggested.  Removed a lot of unnecessary complication.